### PR TITLE
Enable sampling override using thread.id and thread.name

### PR DESCRIPTION
**Pull request by heyams**
fix https://github.com/microsoft/ApplicationInsights-Java/issues/3145

### CODE REVIEW

File: smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SamplingOverridesThreadIdAndName.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+abstract class SamplingOverridesThreadIdAndName {
+
+  @RegisterExtension
+  static final SmokeTestExtension testing =
+      SmokeTestExtension.builder().setSelfDiagnosticsLevel("debug").build();
+
+  @Test
+  @TargetUri(value = "/threadIdAndName", callCount = 100)
+  void testSampling() throws Exception {
+//    assertThat(testing.mockedIngestion.getCountForType("RequestData")).isZero();
```
Comment (heyams) [inline smoke-tests/apps/SamplingOverrides/src/smokeTest/java/com/microsoft/applicationinsights/smoketest/SamplingOverridesThreadIdAndName.java:30]: request sampling override failed. trace works as expected.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/AiSampler.java [hunk 2]
```diff
@@ -57,6 +57,13 @@ public SamplingResult shouldSample(
       Attributes attributes,
       List<LinkData> parentLinks) {
 
+    Attributes extraAttributes =
```
Comment (jeanbisutti) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/AiSampler.java:60]: `threadAttributes` local variable name may be more explicit?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java [hunk 3]
```diff
@@ -83,6 +83,16 @@ private boolean matches(Attributes attributes, @Nullable LazyHttpUrl lazyHttpUrl
       return true;
     }
 
+    static String getValue(Attributes attributes, AttributeKey<String> attributeKey) {
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java:86]:
```suggestion
    static String getValueIncludingThreadAttributes(Attributes attributes, AttributeKey<String> attributeKey) {
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java [hunk 4]
```diff
@@ -121,7 +131,7 @@ private StrictMatcher(String key, String value) {
 
     @Override
     public boolean test(Attributes attributes, LazyHttpUrl lazyHttpUrl) {
-      String val = attributes.get(key);
+      String val = MatcherGroup.getValue(attributes, AttributeKey.stringKey(key.getKey()));
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java:134]:
```suggestion
      String val = MatcherGroup.getValue(attributes, key);
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java [hunk 5]
```diff
@@ -156,7 +166,7 @@ private RegexpMatcher(String key, String value) {
 
     @Override
     public boolean test(Attributes attributes, @Nullable LazyHttpUrl lazyHttpUrl) {
-      String val = attributes.get(key);
+      String val = MatcherGroup.getValue(attributes, AttributeKey.stringKey(key.getKey()));
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java:169]:
```suggestion
      String val = MatcherGroup.getValue(attributes, key);
```

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java [hunk 6]
```diff
@@ -199,7 +209,7 @@ private KeyOnlyMatcher(String key) {
...
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java:212]:
```suggestion
      String val = MatcherGroup.getValue(attributes, key);
```

File: smoke-tests/apps/SamplingOverrides/src/smokeTest/resources/applicationinsights-thread-id-and-name.json [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        {
+          "telemetryType": "trace",
+          "attributes": [
+            {
+              "key": "thread.name",
+              "value": "main",
+              "matchType": "strict"
+            }
+          ],
+          "percentage": 0
```
Comment (trask) [inline smoke-tests/apps/SamplingOverrides/src/smokeTest/resources/applicationinsights-thread-id-and-name.json:81]: I'm not sure this one is needed?
```suggestion
```

File: smoke-tests/apps/SamplingOverrides/src/smokeTest/resources/applicationinsights-thread-id-and-name.json [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+              "matchType": "regexp"
+            },
+            {
+              "key": "http.target",
+              "value": "/SamplingOverrides/thread-id-and-name",
+              "matchType": "strict"
+            }
+          ],
+          "percentage": 0
+        },
```
Comment (trask) [inline smoke-tests/apps/SamplingOverrides/src/smokeTest/resources/applicationinsights-thread-id-and-name.json:60]: I think this one can be removed in favor of the one below

**Comment by trask** [comment]
@heyams I just realized `thread.id` is an `AttributeKey<Long>`. I believe we only support `String` and `List<String>` attribute types in sampling overries. Probably we can just remove `thread.id` support from this PR. I don't imagine that would be something someone would want to sample based on anyways.

File: smoke-tests/apps/SamplingOverrides/src/smokeTest/resources/applicationinsights-thread-id-and-name.json (hunk 8 above)
Comment (heyams) [reply smoke-tests/apps/SamplingOverrides/src/smokeTest/resources/applicationinsights-thread-id-and-name.json:60]: the one below is trace

Comment (trask) [reply smoke-tests/apps/SamplingOverrides/src/smokeTest/resources/applicationinsights-thread-id-and-name.json:60]: oh, i missed that, thx!

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java [hunk 9]
```diff
@@ -83,6 +83,15 @@ private boolean matches(Attributes attributes, @Nullable LazyHttpUrl lazyHttpUrl
       return true;
     }
 
+    static String getValueIncludingThreadAttributes(
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/sampling/SamplingOverrides.java:86]:
```suggestion
    static String getValueIncludingThreadName(
```

**Review by trask — APPROVED** [review]

**Pull request merged**